### ISSUE TITLE
Not working with Unifi Dream Machine Pro and Unifi OS network controller?

### ISSUE THREAD

**Issue by samwelnella** [oznu/homebridge-unifi-occupancy-sensor#20]
**Describe Your Problem:**
I upgraded my UDM Pro to the new Unifi OS and now homebridge-unifi-occupancy-sensor can’t connect to the Unifi network controller. Is the network controller on a UDM Pro supported?

**Logs:**

```
[2/3/2020, 5:49:10 PM] [D Occupancy Sensor] ERROR: Failed to check occupancy: Cannot read property 'length' of undefined
[2/3/2020, 5:49:10 PM] [X Occupancy Sensor] ERROR: Failed to check occupancy: Cannot read property 'length' of undefined
[2/3/2020, 5:49:10 PM] [Anyone Occupancy Sensor] ERROR: Failed to check occupancy: Cannot read property 'length' of undefined
[2/3/2020, 5:49:10 PM] [X Occupancy Sensor] UniFi Events: Connected (1) - wss://172.16.0.1/wss/s/default/events
[2/3/2020, 5:49:10 PM] [PowerView] Hub: Home
[2/3/2020, 5:49:10 PM] [X Occupancy Sensor] UniFi Events: Disconnected (1) - Retry in 5000ms
[2/3/2020, 5:49:10 PM] [Anyone Occupancy Sensor] UniFi Events: Connected (1) - wss://172.16.0.1/wss/s/default/events
[2/3/2020, 5:49:10 PM] [Anyone Occupancy Sensor] UniFi Events: Disconnected (1) - Retry in 5000ms
[2/3/2020, 5:49:10 PM] [D Occupancy Sensor] UniFi Events: Connected (1) - wss://172.16.0.1/wss/s/default/events
[2/3/2020, 5:49:10 PM] [D Occupancy Sensor] UniFi Events: Disconnected (1) - Retry in 5000ms
```

**Homebridge Config:**

```json
{
    "bridge": {
        "name": "Homebridge",
        "username": "XX:XX:XX:XX:XX:XX",
        "port": 51826,
        "pin": "XXX-XX-XXX"
    },
    "description": "This is an example configuration file. You can use this as a template for creating your own configuration file containing devices you actually own.",
    "accessories": [
        {
            "name": "D Occupancy Sensor",
            "unifi": {
                "controller": "https://172.16.0.1/network",
                "username": "XXX",
                "password": "XXX",
                "site": "default",
                "secure": false
            },
            "watch": [
                "XX:XX:XX:XX:XX:XX"
            ],
            "mode": "any",
            "watchGuests": true,
            "interval": 1800,
            "accessory": "UniFi Occupancy Sensor"
        },
        {
            "name": "X Occupancy Sensor",
            "unifi": {
                "controller": "https://172.16.0.1/network",
                "username": "XXX",
                "password": "XXX",
                "site": "default",
                "secure": false
            },
            "watch": [
                "XX:XX:XX:XX:XX:XX"
            ],
            "mode": "any",
            "watchGuests": true,
            "interval": 1800,
            "accessory": "UniFi Occupancy Sensor"
        },
        {
            "name": "Anyone Occupancy Sensor",
            "unifi": {
                "controller": "https://172.16.0.1/network",
                "username": "XXX",
                "password": "XXX",
                "site": "default",
                "secure": false
            },
            "watch": [
                "XX:XX:XX:XX:XX:XX",
                "XX:XX:XX:XX:XX:XX"
            ],
            "mode": "any",
            "watchGuests": true,
            "interval": 1800,
            "accessory": "UniFi Occupancy Sensor"
        }
    ],
    "platforms": [
    ]
}
```

**Screenshots:**

**Environment:**

* **Node.js Version**: 12.14.1
* **NPM Version**: 6.13.4
* **Homebridge Version**: 0.4.50
* **Homebridge Plugin Version**: 1.0.0
* **Homebridge Config UI X Plugin Version**: 4.7.0
* **Operating System**: Docker

**Comment by josiasmontag** [comment]
[This patch](https://github.com/josiasmontag/unifi-events/commit/c3cf0a2b03482e0bb5c9aab8e6ecb83dad71f992) makes it work with the UDM pro.
It breaks support for the older controllers though.

**Comment by hfagelnour** [comment]
thanks for trying to help .... i used the index.js file you proposed and i get the following errorLoaded plugin: homebridge-unifi-occupancy-sensor

/usr/lib/node_modules/homebridge/lib/server.js:171
      plugin.initializer(this._api);
             ^
TypeError: Class constructor UnifiEvents cannot be invoked without 'new'
    at Server.<anonymous> (/usr/lib/node_modules/homebridge/lib/server.js:171:14)
    at Array.forEach (<anonymous>)
    at Server._loadPlugins (/usr/lib/node_modules/homebridge/lib/server.js:145:22)
    at new Server (/usr/lib/node_modules/homebridge/lib/server.js:57:24)
    at module.exports (/usr/lib/node_modules/homebridge/lib/cli.js:32:16)
    at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (internal/modules/cjs/loader.js:1156:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1176:10)
    at Module.load (internal/modules/cjs/loader.js:1000:32)
    at Function.Module._load (internal/modules/cjs/loader.js:899:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)
    at internal/main/run_main_module.js:18:47
[4/19/2020, 5:31:39 AM] [HB Supervisor] Homebridge Process Ended. Code: 1, Signal: null
[4/19/2020, 5:31:41 AM] [Config] Homebridge restart request received

**Comment by hfagelnour** [comment]
@josiasmontag how to use this patch?please

**Comment by josiasmontag** [comment]
Yo need to manually replace `node_modules/homebridge-unifi-occupancy-sensor/node_modules/unifi-events/index.js` with my patched file.

**Comment by hfagelnour** [comment]
@josiasmontag  OUTSTANDING..... this worked like a charm..... many thanks
I wonder if such a patch is available to get the UDMP unifi protect and its events work on the homebridge using the plugin "Homebridge Unifi Protect Camera Motion"

**Comment by hfagelnour** [comment]
After monitoring the homebridge readings and its HomeKit occupancy reporting it for some odd reason doesn't update the status as it should.
the homebridge log shows it is succeeding in probing the UDMP as it should and getting back responses but still all the responses basically shows the probed clients as available even if they switch off the wifi or leave the premises.
any thoughts welcomed, pls

**Comment by JDogg016** [comment]
I see there is a patch to get this plugin to work with UDM pro. Question: how do you replace the index.js file if you are running homebridge on a docket container. 

Also (simpler) would someone post the entire plugin as specific to UDM Pro?

**Comment by oznu** [comment]
Should be resolved in v2.0.0.

https://github.com/oznu/homebridge-unifi-occupancy-sensor/pull/28
